Extend CiliumNetworkPolicy validation to allow IPv6 CIDRs

joestringer opened the pull request:
On Kubernetes versions 1.9 and later, we validate the contents of CiliumNetworkPolicy resources when they are applied to the kubernetes cluster. Previously, the regex for validating CIDR policies would only allow IPv4 prefixes, so if a user attempts to install a CIDR policy that allows IPv6 traffic, the validator would reject it with an error like:

```
validation failure list:
must validate one and only one schema (oneOf)
spec.egress.toCIDR in body should match '^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/([0-9]|[1-2][0-9]|3[0-2])$'
```

This PR loosens the restrictions on CIDR policies to allow IPv6 addresses to be specified as well.

Note that if you deploy CIDR IPv6 policies, then attempt to downgrade to a version that does not contain this patch, then Cilium will go into a bad state because it is unable to apply your previous policy. The minimum expected versions when using CIDR IPv6 policies are `v1.0.2` or `v1.1`. 

Fixes: #4000

joestringer committed:
k8s: Support IPv6 addresses in CIDR policy

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
k8s: Add CRD IP address validation unit tests

Signed-off-by: Joe Stringer <joe@covalent.io>

raybejjani commented on pkg/k8s/apis/cilium.io/v2/register_test.go:
```diff
@@ -77,3 +78,29 @@ func (s *CiliumV2RegisterSuite) TestNeedsUpdateCorruptedVersion(c *C) {
+		}
+		// The below is always false, we shouldn't hit this with valid
+		// CIDR prefixes.
+		c.Assert(input, Equals, "failed to match CIDR.OneOf[*].Pattern")
```
This assert is fatal to the test, right? From the code structure, it looks like you intended to continue the outer loop, though.

joestringer replied:
That's fine, these are all positive tests so if this ever triggers, then something is wrong. I manually added some counter-examples in local testing and observed that this causes the test to fail.

A subsequent update could extend this testsuite to add some other negative examples.

joestringer replied:
I added negative tests.

raybejjani commented on pkg/k8s/apis/cilium.io/v2/register_test.go:
```diff
@@ -77,3 +78,29 @@ func (s *CiliumV2RegisterSuite) TestNeedsUpdateCorruptedVersion(c *C) {
 }
+
+func (s *CiliumV2RegisterSuite) TestCIDRRegex(c *C) {
+	testCIDRs := []string{
```
Since the regex is inscrutable, is there any chance we can add some more IPv6 addresses here? in particular, longer prefixes and ones with the middle 0s filled in (I'm pretty sure that is legal and the `::` is just a convenience).

joestringer replied:
Sure, I added a few examples. Let me know if you have any more that you'd like to test.

raybejjani replied:
I think that covers it. I just didn't want to miss a more verbose address. thanks!

raybejjani reviewed (approved):
This seems fine to me, but I added aanm in case there's something I missed

joestringer committed:
docs: Describe downgrade impact of IPv6 CRD validation

Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -359,25 +359,25 @@ var (
+					`{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}` +
+					`))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]` +
+					`{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d))` +
+					`{3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:` +
```
The alternative on this line and the next is allowing the dot notation for IPv4-mapped IPv6 addresses.

rlenglet commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -359,25 +359,25 @@ var (
+					`{3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:` +
+					`[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|` +
+					`1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|` +
+					`((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d` +
```
This one too.

joestringer replied:
I can't figure out why the regex isn't matching IPv4-mapped IPv6 addresses, but I can deconstruct the regex and strip out the IPv4 references.

rlenglet reviewed (dismissed):
The regexp still allows IPv4-mapped IPv6 addresses: https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
Those shouldn't be allowed, as I'm not sure we're handling those properly.

joestringer committed:
k8s: CIDR: Expand v6 regex to make it more readable

No functional changes.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
k8s: CIDR: Disallow IPv4-mapped IPv6 addresses

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
k8s: CIDR: Format IPv6 CIDR regex

The precheck script complains if you don't consistently indent things,
appease it even if it makes the regex harder to read.

Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on pkg/k8s/apis/cilium.io/v2/register_test.go:
```diff
@@ -16,6 +16,7 @@ package v2
 
 import (
 	. "gopkg.in/check.v1"
+	"regexp"
```
We usually change the order of imports, and always group the stdlib imports first.
